### DISCUSSION TITLE
the real problem with IKEMEN.........  ideas and dreams of the future

### DISCUSSION THREAD

**Discussion by ZyonZZ** [ikemen-engine/Ikemen-GO#1904] [Feature Request]
### Is your feature request related to a problem? Please describe.

The rise of fighting games was 100% in the Arcades, in the 90s they had a huge boom, where even the same companies released a fighting game every year and it was profitable (king of fighters) we had a large number of games in those times from the Legendary SF2 to street fighter 3rd strike which was later the generational leap to 3D, and when the arcades were already in decline... now for the new generations, the learning curve that fighting games have has made few They have had the resounding success that they had in their golden age...

### Describe the solution you'd like

Currently, in 2024, common users who only want to play a quick game have very little programming experience... a LAUNCHER is necessary that facilitates playing with people and the integration of the common user with IKEMEN... and that points directly to the Online
What would this be like and what would be its use?

Have you ever played Fightcade2?? is a launcher that allows you to play classic arcade games online like kof2k2, mvc, sf, etc.

Now the common User just has to download the launcher, select the rom, configure controls and that's it, he can challenge anyone in that game who is in the Lobby.

For those who don't know it, go watch a video of what it is like... and all the advantages it has... and with that in mind we continue... with ikemen.

in ikemen..
A launcher like that would be necessary in ikemen
This launcher would be executable... it would open a window... there it could have information about patches, improvements, engine updates and so on (that the launcher updates automatically)
the fullgames that you have selected..
and gives you the option to open the game or enter an online Lobby

Now we would also need ANOTHER application that only the ikemen launcher can use...
The function of this is to package the games... with a release code...
This application's function is that the user who is not the creator of the fullgame... does not modify said characters and each ikemen game has its version, the function of this is to avoid desynchronizations between users and their cheating.

I have this in mind...
You enter the Lobby... there you see a list of all the users who have the ikemen launcher open... with their respective pings...

then you open a "party room" (provisional name)
you coming in as a leader...

You select from your Ikemen games the one you want to play, let's say for this example ""Marvel Vs Capcom 5 ultimate revenge ikemen ver2.1""
The name of the game would appear next to your room in the lobby... so that others in the lobby can see that you have that complete game and that they can enter.

Within the ""party room"" options are...
the number of players per room and observers
(watchers have lower priority when downloading or launching the game)
room password: (for if you want to play exclusively with friends)
kickplayer. ping limit:

Cooperative history
**2vsAI** **3vsAI** **Tagbattle**
Debug buttons on/off (this option only appears if you select History coop)
Imagine being able to finish a story mode in such a way
coop 2players
coop 3players
in a kof one character per player
make combinations in tagbattle each one a character..
AI handicap (This modifies the life of the AI ​​chars because modifying mobility is difficult... it could be a bar that increases life, normal, or lowers life basically the stats of the char only))

VS competitive **1vs1** **2vs2players** **3vs3players.**

When the user enters your "party room" depending on what you chose, he can position himself on the side of your team or VS or observer.
In addition to starting to download, basic data of the game ""Marvel vs Capcom 5 ur"" only screenpack, fight and music from the select...

For the others who are inside the ""party room"" a Ready button would appear (and below the button a bar that says downloading basic elements something like this and the button in gray)(((basic elements are screenpack, fight and music from the vs and character select)))
I would download this data from the Leader of the ""party room""

Everyone can see the room configuration, the game mode selected by the leader and the ping of each player.

and when everyone has downloaded the basics...
You can click Launch Game... and it sends you directly to the Character Selector... because the modes were previously chosen from the party room options

to play online with other people.
It is not necessary for everyone to have the full game.
for a fight...
You would need the chars, the stage, the screenpack system, the fight (life bars, names and so on) and music... and all that weighs a few megabytes (except some chars)
and that's it.. if you don't have that data.. 

on the vs. screen.. ikemen would download it to the rival or ally.. (and thus he gradually obtains the game by downloading what is necessary.. easier if they are users that they like the same character.. less download would be needed)

The "vs screen" is where you would download... chosen characters and stages...
you don't need to download all characters at once..

If the users liked said game, they could download the FULLGAME separately and paste it in their "packaged" folder so that ikemen could see it.
and reduce download time to 0

Another option that could be added to the room is::
Arcade VS...
This is the classic mode where one can be playing in arcade mode... and advancing in the story...
someone enters your room and challenges you... this mode only accepts 1vs1

This could be exclusive to the fullgames, for the practicality of arriving and challenging without downloads... in a game already in progress... by synchronizing it.

I know....this reads like it's impossible...and maybe there are things I'm leaving out...I'm just saying that
Ikemen is a good engine, and online it is not used in a good way at all.
If those from fightcade2 could do something like that
I would even accept that said launcher had simple advertising banners so that they would also generate some support for the creator of the engine..
and in the "party rooms" there are even banners supporting full games, something that would boost the creators of full games every time someone plays their creation.

even an integration of all this for android... but that's already uff... too complicated... just dreams

All of this occurred to me because there are many players who are interested in these fighting games, but don't know how to start...
several times I have seen people on tiktok lives asking how to download the game or if it is for android....and really...for a common user...it is not easy at all to enter the world of ikemen...if you are completely new and You just want to play some challenges with friends or someone.

**Comment by Kiwi1444** [comment]
The entry threshold for Ikemen Go could be lower if the launcher allowed adding new characters and stages. Only imagine: you drag & drops archive into laucher window and only choose .def file you want to install. The launcher does everything else itself.

**Reply by ZyonZZ** [reply]
My opinion and suggestion is based on what I have seen on mugen's tiktok lives...
And as I told you... I have not seen people say: ""or look, you can create a game with this, I want to create one""
People already want the game already created, that's why the comments are...

What is the ROM to download??? What is the name of the game? is it for android?
100% of the comments are the same..

I have told you... it is an engine to create fighting games... put ikemen in the search engine and it will come out...
People don't want to learn how to do it, they already want things done...

For that reason we are not going to educate people.
You see a darksouls game, Zelda or any franchise... and you tell people... "Hey, do you want a game like that??? go program it""

NO...
People want to play it, the experience of having it...
to those people as their first contact with ikemen... make the game easier for them...
Now if any of these want to investigate more, they can do it...

mugen/ikemen is fantastic because it gives you the experience that other companies cannot give you... you can play Superman vs Spiderman vs Goku vs iori yagami...
It makes things real that the industry would have to invest a lot of licenses and money in.
For that reason it is easy to attract attention.
but making the process difficult does not help much.


For these reasons I consider that for the moment we let the programmers take care of the Fullgames...
and users the simplest process so they can play
not only so that the ikemen community is larger but so that there is a greater probability that more people will join this.

**Reply by Kiwi1444** [reply]
There are already a lot of complete game builds for Mugen, but their legal status is unclear.
As for me Mugen always was character-oriented platform. Something like fighting game constructor with all the attendant pros and cons. With a unique opportunity to put characters from all game universes face to face and find out who is the strongest: legendary General from Kaiser knuckle or some custom build of Akuma?

**Reply by ZyonZZ** [reply]
yes yes, exactly... and IKEMEN has the possibility of making compatible ALLL THOSE FULLGAMES... that are already built in mugen...
and with the IKEMEN launcher that I propose to facilitate ONLINE... for the common user.... This is how I see the future of ikemen.

**Comment by formula1** [comment]
Some related thoughts


# Problem: Too Many Characters
For a given game, there may come a point when there are more legal characters than a user is interested in downloading. Currently there a character select screens that have so many characters that it's hard to see individual characters.
![too-many-characters](https://github.com/user-attachments/assets/e3002d75-07ba-45f5-af8d-553848016d52)

## Solution: Download On Demand
In this case, instead of players downloading all the characters that the opponent has, after each player selects their character, they download eachothers and load them into Ikemen.

After each match, they can choose to keep their opponents character if they liked them or not. There probably will be some troubles to avoid saving duplicates, but creating a sha of each character's code and checking file sizes there might be a clean and quick way to check for that.


# Problem: Too Many Characters for Prepackaged Games
An argument is that for prepackaged games, too many characters won't be a problem. But a quick check on the total number of street fighter characters amounts to 51. And I expect capcom will be interested in creating even more new characters in the future. As a result, expecting prepackaged games to have too many characters may eventually become a reality.

## Solution: Character White List
While a prepackaged game can have potentially 50+ characters for players to choose from. They can also just provide a "whitelist". These are all the legal characters that can be used in their prepackaged games. This way players don't need to download the entire character list but instead download only the characters they are interested (for example: only S and A tier characters) and they can still play online and download characters they dont care about on demand.

If a player tries to use a character that isn't on the whitelist, the other player would
- download the player
- create a sha from download, see if it exists on the whitelist
- check the file size, see if it matches on the whitelist
- check info like names and etc, see if it matches on the whitelist

If any of the above fails, the match fails to start

# General Issues
- Limit Character Sizes
- Code that crashes games

**Reply by ZyonZZ** [reply]
Maybe in my original post I didn't make myself understood...

and I'm not sure which idea is more feasible, the one you propose... or the one I propose in the original post...

<blockquote>

"<i>For the others who are inside the ""party room"" a Ready button would appear (and below the button a bar that says downloading basic elements something like this and the button in gray)(((basic elements are screenpack, fight and music from the vs and character select)))
I would download this data from the Leader of the ""party room""

Everyone can see the room configuration, the game mode selected by the leader and the ping of each player.

and when everyone has downloaded the basics...
You can click Launch Game... and it sends you directly to the Character Selector... because the modes were previously chosen from the party room options

to play online with other people.
It is not necessary for everyone to have the full game.
for a fight...
You would need the chars, the stage, the screenpack system, the fight (life bars, names and so on) and music... and all that weighs a few megabytes (except some chars)
and that's it.. if you don't have that data.. on the vs. screen.. ikemen would download it to the rival or ally.. (and thus he gradually obtains the game by downloading what is necessary.. easier if they are users that they like the same character.. less download would be needed)

If the users liked said game, they could download the FULLGAME separately and paste it in their "packaged" folder so that ikemen could see it.
and reduce download time to 0 </i>" </blockquote>


Here I propose that... the basics be downloaded when entering the ""party room""
the basic thing is.. screenpack, fight and music from the vs and character select..
Now when the players directly enter the "Characterselect" they will choose their characters.

The "vs screen" is where you would download... chosen characters and stages...
you don't need to download 51 characters at once..
but maximum 6.. ((If there are 6 players in the room))

or if you play 1vs1.. download only 2 chars on the vs. screen.
And if options like "Remach" would speed things up... the stages and music normally don't weigh much...

Maybe I don't know which one is better... programmers could even choose another method where this could be streamlined...
So I don't know what the best method is.

**Reply by formula1** [reply]
@ZyonZZ 

https://github.com/ikemen-engine/Ikemen-GO/wiki/Miscellaneous-info#general-information

I think yours might be more doable considering the state of the current ikemengo online. Just figured I would put in my 2 cents since I think a lot of people want to mess around with lots of characters.

- Mugen Packs are uploaded to a web accessible place
  - Mugen Packs include screenpack, fight and music from the vs and character select..
- People can create/join room of a Mugen Pack
- When everyone is ready, Game starts as normal using the Room's Mugen Pack
